src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -100,12 +129,49 @@ public async Task AuthenticateAsync(IConnection connection, ConnectionDescriptio
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
-            using (var conversation = new SaslConversation(description.ConnectionId))
+            var firstStepResult = description.IsMasterResult.SpeculativeAuthenticate;
+
+            if (firstStepResult == null)
+            {
+                await AuthenticateAsync(
+                        connection, description, previousStep: null, previousStepResult: null, cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                var (firstStep, _) = CreateFirstStepAndCommand(_mechanism, connection, conversation: null, description);
+                await AuthenticateAsync(
+                    connection,
+                    description,
+                    previousStep: firstStep,
+                    previousStepResult: firstStepResult,
+                    cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async Task AuthenticateAsync(
```

Comment:
An alternative name for this could be `ResumeAuthenticateAsync`

Reply:
Deferring discussion to non-outdated-thread here:  https://github.com/vincentkam/mongo-csharp-driver/pull/95/files#r402608242